### ISSUE TITLE
Your service workers are throwing errors in safari.

### ISSUE THREAD

**Issue by brainkim** [thepracticaldev/dev.to#5434]
**Describe the bug**

Unable to load the site in safari on repeat visits.

**To Reproduce**

Open the website in safari. Maybe wait a day or something I have no idea.

**Expected behavior**

The website loads.

**Screenshots**
![Screen Shot 2020-01-10 at 5 51 51 AM](https://user-images.githubusercontent.com/2996754/72148085-f9958900-336d-11ea-96db-211002b2b82b.png)
![Screen Shot 2020-01-10 at 5 53 23 AM](https://user-images.githubusercontent.com/2996754/72148067-f00c2100-336d-11ea-90d4-e3ce53f5f4f7.png)
![Screen Shot 2020-01-10 at 5 54 28 AM](https://user-images.githubusercontent.com/2996754/72148042-dff44180-336d-11ea-8ad3-2bfb3d48c37f.png)

**Desktop (please complete the following information):**

- OS: macOS Mojave 10.14.6
- Browser: Safari 13.0.1

**Additional context**

Service workers, not even once.

**Comment by rhymes** [comment]
Hi @brainkim, I have macOS 10.14.6 but oddly I have Safari 13.0.4 and the website loads.

![Screenshot 2020-01-10 at 1 05 40 PM](https://user-images.githubusercontent.com/146201/72151968-f28c6c00-33a9-11ea-9cf1-c7bf80cb0e19.png)


Do you have any particular extensions?

**Comment by brainkim** [comment]
I'm running adguard but it would be a pain to turn off and I doubt that adguard is the issue. This issue only appears after maybe a day or so, meaning I think some cache item is being evicted that a dev.to dev does not expect. I visit the site, see dev.to stuff, go about my merry way, and then visit the site again only to have loading stuck and the site completely unable to load until I clear the site’s service worker.

I would take a closer look at the error messages I screenshotted. It looks like  you haven’t sufficiently guarded against a cache being empty:

https://github.com/thepracticaldev/dev.to/blob/482a77a3753525f1058facd5fd1af016a6ec96c9/app/views/service_worker/index.js.erb#L40

This line does a null check but it will always return true because cache.match returns a promise, right?

**Comment by brainkim** [comment]
Did you deploy a service worker update today? I’m suddenly able to visit the site again but expect this issue to pop up again soon. (I was intentionally not clearing the service worker so that I’d be able to help debug)

**Comment by brainkim** [comment]
It’s broken again as I check today.

**Comment by brettimus** [comment]
Still broken for me in Safari 13.0.5 -- I can't reproduce the exact conditions under which this happens, but I saw the same errors that @brainkim highlighted - where it appears the response is undefined 

The only extension I have is 1Password.

Also, the workaround for me appears to be opening a tab in private-mode, reloading dev.to, then trying again in not-private mode.

**Comment by brettimus** [comment]
> I'm running adguard but it would be a pain to turn off and I doubt that adguard is the issue. This issue only appears after maybe a day or so, meaning I think some cache item is being evicted that a dev.to dev does not expect. I visit the site, see dev.to stuff, go about my merry way, and then visit the site again only to have the site fail to load until I clear the site’s service worker.
> 
> I would take a closer look at the error messages I screenshotted. It looks like you haven’t sufficiently guarded against a cache miss:
> 
> https://github.com/thepracticaldev/dev.to/blob/482a77a3753525f1058facd5fd1af016a6ec96c9/app/views/service_worker/index.js.erb#L40
> 
> This line does a check for a match but the expression will always be false because `cache.match` returns a promise, right?

BTW  [according to MDN](https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage/match), `caches` will return `undefined` when there's no match. 

> This method returns a Promise for a Response , or undefined if no match is found.

**Comment by maestromac** [comment]
I use Safari 13.0.5 (with bitwarden, adguard, and grammarly) as my default and I can access DEV without issue. 

@brettimus are you accessing DEV as logged in user or logged out? and how often are you running into this issue?

**Comment by brettimus** [comment]
> I use Safari 13.0.5 (with bitwarden, adguard, and grammarly) as my default and I can access DEV without issue.
> 
> @brettimus are you accessing DEV as logged in user or logged out? and how often are you running into this issue?

Hey @maestromac! Thanks for replying.

I've hit this issue two or three times so far this year, and it usually happens if I was logged in previously. It doesn't seem to happen when I don't have an active (or expired?) session, as I'm always able to get to the homepage in private mode.

**Comment by brainkim** [comment]
@brettimus 
> BTW according to MDN, caches will return undefined when there's no match.

This looks like an unfortunate case of an ambiguous modifier. The `caches.match` method will always return a promise. See the docs on the return value, or the [actual spec](https://w3c.github.io/ServiceWorker/#dom-cachestorage-match).

**Comment by brettimus** [comment]
> @brettimus
> 
> > BTW according to MDN, caches will return undefined when there's no match.
> 
> This looks like an unfortunate case of an ambiguous modifier. The `caches.match` method will always return a promise. See the docs on the return value, or the [actual spec](https://w3c.github.io/ServiceWorker/#dom-cachestorage-match).

Wow you're totally correct! Just re-read and it says it resolves with `undefined` 🤦🏻‍♂️ And I'm glad you're right -- I was concerned that a browser api was mixing async/sync

This means you're absolutely correct! .

**That's gotta be what's giving us trouble, since the initial fetch is against `/shell_top`**, which _should_ be short-circuited by that if-condition... but never will be!

https://github.com/thepracticaldev/dev.to/blob/482a77a3753525f1058facd5fd1af016a6ec96c9/app/views/service_worker/index.js.erb#L47

**Comment by brettimus** [comment]
just hit this again. will look into creating an MR given I have a lot more at-home time these days 😬 

<img width="1098" alt="image" src="https://user-images.githubusercontent.com/3201327/77455698-7f1bc780-6dfa-11ea-82ad-0186c4f143ba.png">

**Comment by brettimus** [comment]
posting again just to keep myself accountable -- forked and added (what I assume to be) a fix today.

I've been spending my lunch break getting the dev environment set up (lots of waiting for docker right now 😁)

will hopefully open a PR and discuss how to write a test for this if possible

**Comment by rhymes** [comment]
Awesome @brettimus, thanks a lot!

**Comment by brettimus** [comment]
btw opened the PR last night!

**Comment by brettimus** [comment]
hit this again and it sent me down a rabbit hole of how to unregister a service worker within a service worker in safari dev tools...

admittedly the rabbit hole was not that deep. just got execute some code in the console of the service worker (it's the only way! I forgot how frustrating safari dev tools can be given that you're not allowed to open the inspector until you're on a fully-loaded page)

anyways, this is another workaround for when the page freezes due to the service worker error:

```js
  navigator.serviceWorker.getRegistrations().then(rs => rs[0].unregister)
```

**Comment by brettimus** [comment]
another update:

i hit the bug again today. 

i had an idea on how to reproduce it: by executing `caches.delete` with both `shell_top` and `shell_bottom` as args, in the main console as well as the js console... it did not reproduce the bug! so strange.

**Comment by brettimus** [comment]
so we fixed one bug, and another appeared! (it has same consequence: the site does not load in safari, so this bug should remain open sadly)

<img width="850" alt="image" src="https://user-images.githubusercontent.com/3201327/82199253-be97f900-98fd-11ea-967c-3254eb655e45.png">


<img width="828" alt="image" src="https://user-images.githubusercontent.com/3201327/82199229-b50e9100-98fd-11ea-96f6-7cd653d8dc2b.png">

**Comment by benhalpern** [comment]
This seems fixable! If the issue is that `cachedResponse` is `undefined` then we can guard statement this area or even do `(cachedResponse || {}) =>` or something like that. `{}['version']` will in fact not equal `json['version']`.

**Comment by brainkim** [comment]
Maybe a lil bit of TypeScript might catch bugs like this as well? 😉

**Comment by atsmith813** [comment]
This seems to have been resolved with #9250 🎉
